### PR TITLE
Implement dilation POVMs

### LINKED ISSUE #66: Implement dilation POVMs

### PR DESCRIPTION
**Pull request by timmintam**
This PR aims to provide an implementation of dilation POVMs, which will be added to the `povm_toolbox.library`.
Closes #66.

### CODE REVIEW

File: povm_toolbox/library/dilation_measurement.py [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+
+        This method takes a supplied circuit and appends the measurement circuit to it.
+        An ancilla qubit is required for each qubit that we want to measure (specified by
+        :attr:`.measurement_layout`). If the input circuit has one or more idling qubit(s),
+        they will be used as measurement ancilla qubits. If more ancilla qubits are needed,
+        those will be added to the input circuit (therefore increasing its size).
+
+        .. warning::
+           The number of qubits in the input circuit may be increased due to the ancilla
+           qubits qubits required for dilatation measurements.
```
Comment (mrossinek) [inline povm_toolbox/library/dilation_measurement.py:187]:
```suggestion
           qubits required for dilatation measurements.
```

File: povm_toolbox/library/dilation_measurement.py [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+    @override
+    def reshape_data_bin(self, data: DataBin) -> DataBin:
+        return data
+
+    @override
+    def _povm_outcomes(
+        self,
+        bit_array: BitArray,
+        povm_metadata: POVMMetadata,
+        loc: int | tuple[int, ...] | None = None,
```
Comment (mrossinek) [inline povm_toolbox/library/dilation_measurement.py:232]:
```suggestion
        povm_metadata: POVMMetadata,
        *,
        loc: int | tuple[int, ...] | None = None,
```
Can you please propagate this also to all other `_povm_outcomes`  methods in the other classes?

File: povm_toolbox/library/dilation_measurement.py [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+        bitstring_outcome: str,
+    ) -> tuple[int, ...]:
+        """Transform a bitstring outcome to a POVM outcome.
+
+        Args:
+            bitstring_outcome: the raw outcome of the measurement. The order of
+            qubit is assumed to be reversed. More specifically, the first
+            :attr:`.num_qubits` bits will correspond to the ancilla qubit (in
+            reverse order) and the last :attr:`.num_qubits` bits will correspond
+            to the qubit to measure (in reverse order).
```
Comment (mrossinek) [inline povm_toolbox/library/dilation_measurement.py:260]:
```suggestion
            bitstring_outcome: the raw outcome of the measurement. The order of
                qubit is assumed to be reversed. More specifically, the first
                :attr:`.num_qubits` bits will correspond to the ancilla qubit (in
                reverse order) and the last :attr:`.num_qubits` bits will correspond
                to the qubit to measure (in reverse order).
```

File: povm_toolbox/library/dilation_measurement.py [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+        """
+        return tuple(
+            2 * int(bit_q) + int(bit_a)
+            for bit_q, bit_a in zip(
+                bitstring_outcome[: self.num_qubits - 1 : -1],
+                bitstring_outcome[self.num_qubits - 1 :: -1],
+            )
+        )
+
+    def _unitary_from_parameters(self, param: np.ndarray):
```
Comment (mrossinek) [inline povm_toolbox/library/dilation_measurement.py:274]:
```suggestion
    def _unitary_from_parameters(self, param: np.ndarray)  -> np.ndarray:
```

File: povm_toolbox/library/dilation_measurement.py [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+        """Construct the unitary defining the dilation POVM from parameters.
+
+        We use the same parametrization of the dilation POVM as the one presented in the
+        work of G. García-Pérez, M. A. Rossi, B. Sokolov, F. Tacchino, P. K. Barkoutsos,
+        G. Mazzola, I. Tavernelli, and S. Maniscalco, “*Learning to measure: adaptive
+        informationally complete generalized measurements for quantum algorithms*”, PRX
+        Quantum 2, Publisher: American Physical Society, 040342 (2021). Refer to this work
+        for a detailed explanation of the parametrization.
+        """
```
Comment (mrossinek) [inline povm_toolbox/library/dilation_measurement.py:283]: Even though private, please also add the `Args` and `Returns` sections to this docstring

File: povm_toolbox/library/dilation_measurement.py [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+
+    def _unitary_from_parameters(self, param: np.ndarray):
+        """Construct the unitary defining the dilation POVM from parameters.
+
+        We use the same parametrization of the dilation POVM as the one presented in the
+        work of G. García-Pérez, M. A. Rossi, B. Sokolov, F. Tacchino, P. K. Barkoutsos,
+        G. Mazzola, I. Tavernelli, and S. Maniscalco, “*Learning to measure: adaptive
+        informationally complete generalized measurements for quantum algorithms*”, PRX
+        Quantum 2, Publisher: American Physical Society, 040342 (2021). Refer to this work
+        for a detailed explanation of the parametrization.
```
Comment (mrossinek) [inline povm_toolbox/library/dilation_measurement.py:282]: I don't think we have to duplicate this from the main docstring. You can simply refer to it if you want to add an explicit link.
But leaving that implied from this internal method should be fine, too

File: povm_toolbox/library/povm_implementation.py [hunk 7]
```diff
@@ -171,6 +173,38 @@ def compose_circuits(self, circuit: QuantumCircuit) -> QuantumCircuit:
                 f" {self.num_qubits} qubits."
             )
 
+        # Check if the measurement circuit requires some ancilla qubits
+        if self.measurement_circuit.num_qubits > self.num_qubits:
```
Comment (mrossinek) [inline povm_toolbox/library/povm_implementation.py:177]: Please add a note to this methods' docstring about how it handles this case.

File: povm_toolbox/library/povm_implementation.py [hunk 8]
```diff
@@ -171,6 +173,38 @@ def compose_circuits(self, circuit: QuantumCircuit) -> QuantumCircuit:
...
+            index_layout = copy(index_layout)
+            # Get idle qubits in supplied circuit (to be used as ancilla for measurement circuit)
+            idle_qubits = list(circuit_to_dag(dest_circuit).idle_wires())
+            # Get the indices of the idle qubits
+            idle_index = [dest_circuit.qubits.index(qubit) for qubit in idle_qubits]
+            # Remove the idle qubits that will be measured (as specified by index_layout)
+            idle_index = [idx for idx in idle_index if idx not in index_layout]
+            # TODO: why would someone want to measure an idle qubit ? Should we raise a warning ?
```
Comment (mrossinek) [inline povm_toolbox/library/povm_implementation.py:185]: Let's raise an error when this happens. I'd rather have this not be supported at all then to overwrite a case a user might have triggered specifically.
If a user would like this to work, they can open an issue to request this to be working.

File: test/test_utilities.py [hunk 9]
```diff
@@ -28,3 +33,31 @@ def test_double_ket_to_matrix(self):
...
+                ),
+            )
+        )
+        self.assertTrue(np.allclose(np.linalg.norm(onb, axis=0), np.ones(3)))
+
+    def test_n_sphere(self):
+        """Test the ``n_sphere`` utility function."""
+        angles = np.array([0.3, 0.9, 0.1])
+        unit_vector = n_sphere(angles)
+        print(unit_vector)
```
Comment (mrossinek) [inline test/test_utilities.py:59]:
```suggestion
```

File: povm_toolbox/library/povm_implementation.py (hunk 8 above)
Comment (timmintam) [reply povm_toolbox/library/povm_implementation.py:185]: These lines are visited only if the measurement circuit requires some ancilla qubits. Should we move this check outside this `if` statement (i.e., check/raise all the time ?)

**Review by mrossinek — APPROVED** [review]

**Pull request merged**